### ISSUE TITLE
using stdin no longer works

### ISSUE THREAD

**Issue by jdavidheiser** [zheller/flake8-quotes#30]
Running flake8 with the following syntax

`cat file.py | flake8 -`

does not work with flake8-quotes, but does work with other packages.  This is most likely caused by the fact that recent versions of flake8 are now importing pycodestyle and not pep8.

**Comment by xZise** [comment]
I get the following error:
```
$ cat flake8_quotes/__init__.py | flake8 -
Traceback (most recent call last):
  File "/home/xzise/.pyenv/versions/3.5.2/bin/flake8", line 11, in <module>
    sys.exit(main())
  File "/home/xzise/.pyenv/versions/3.5.2/lib/python3.5/site-packages/flake8/main.py", line 36, in main
    report = flake8_style.check_files()
  File "/home/xzise/.pyenv/versions/3.5.2/lib/python3.5/site-packages/flake8/engine.py", line 181, in check_files
    return self._retry_serial(self._styleguide.check_files, paths=paths)
  File "/home/xzise/.pyenv/versions/3.5.2/lib/python3.5/site-packages/flake8/engine.py", line 172, in _retry_serial
    return func(*args, **kwargs)
  File "/home/xzise/.pyenv/versions/3.5.2/lib/python3.5/site-packages/pycodestyle.py", line 1877, in check_files
    runner(path)
  File "/home/xzise/.pyenv/versions/3.5.2/lib/python3.5/site-packages/flake8/engine.py", line 126, in input_file
    return fchecker.check_all(expected=expected, line_offset=line_offset)
  File "/home/xzise/.pyenv/versions/3.5.2/lib/python3.5/site-packages/pycodestyle.py", line 1608, in check_all
    self.check_ast()
  File "/home/xzise/.pyenv/versions/3.5.2/lib/python3.5/site-packages/pycodestyle.py", line 1555, in check_ast
    for lineno, offset, text, check in checker.run():
  File "/home/xzise/Programms/flake8-quotes/flake8_quotes/__init__.py", line 57, in run
    file_contents = self.get_file_contents()
  File "/home/xzise/Programms/flake8-quotes/flake8_quotes/__init__.py", line 52, in get_file_contents
    return pep8.stdin_get_value().splitlines(True)
  File "/home/xzise/.pyenv/versions/3.5.2/lib/python3.5/site-packages/pep8-1.7.0-py3.5.egg/pep8.py", line 1198, in stdin_get_value
    return TextIOWrapper(sys.stdin.buffer, errors='ignore').read()
ValueError: I/O operation on closed file
```

Using the following version: `2.6.2 (pycodestyle: 2.0.0, flake8_quotes: 0.3.0, mccabe: 0.5.0, pyflakes: 1.2.3) CPython 3.5.2 on Linux`

**Comment by twolfson** [comment]
The comment from @xZise has been determined due to an empty file and not an issue with `flake8` itself. This issue is currently being caused by a change in `flake8's` beta releases which we won't be supporting until https://gitlab.com/pycqa/flake8/issues/161 is fully resolved. Closing this issue for now.

**Comment by xZise** [comment]
I think you got confused: For one this is about `stdin` so there is no “empty file” and `stdin` is not empty here. Also the issue with the empty file is only present when an error is reported on them, which wasn't happen with any of the plugins installed (`flake8-future-import` was the reason for the error). This is also not about Flake8 3.x but also Flake8 2.x. I got the “empty file” error only on 3.x but as you can see, the error is caused by Flake8 2.x.

Mentioning bug 161 is correct but would only show that this plugin is actually incorrectly reading from stdin. But to be honest the documentation is basically non-existent so I have no idea what is actually the proper way and I'm just guessing my way through the bug.

Apparently importing `flake8` before importing `pep8` should work. I haven't tested this an I'm skeptical whether this imports `pycodestyle` instead of `pep8` when installed. There are also plugins like `flake8-coding` which do not import `flake8` and they seem to get by for some reason, although they import `pycodestyle` first fall back to `pep8`. This is basically what I did with importing `flake8.engine.pep8` as that will the pep8 package used by Flake8, independently from `pep8` and `pycodestyle`.

So tl;dr: This bug is not fixed and should be reopened, unless you are certain that reading from stdin works (as in “report errors”) for Flake8 2.x.

**Comment by xZise** [comment]
Okay apart from my traceback above, it doesn't even work currently on older versions of Flake8 2.x using pep8 and not pycodestyle (unlike #31 seems to fix):

```
$ echo 'this_should_be_linted = "double quote string"' | flake8 - ; flake8 --version
[]
2.4.0 (pep8: 1.5.7, flake8_quotes: 0.5.1, mccabe: 0.3, pyflakes: 0.8.1) CPython 2.7.8 on Linux
```

I used master but added a line to output the file contents, which are an empty array.

**Comment by twolfson** [comment]
Sorry, there has been a lot of PRs flying around and it's hard to keep track of which is which. Please update the PR (#38) with a failing/passing test case to prove this regression has been patched and no longer poses an issue.

**Comment by xZise** [comment]
Okay I have done a bit more testing without and with #38:

* Flake8 2.4.1 and 2.5.0 cause the ValueError from above regardless of the patch
* Flake8 2.5.5 work regardless of the patch (2.5.1 has monkey-patched `pep8.stdin_get_value` which I verified: 2.5.1 and 2.5.2 work with the patch but I haven't tested them without the patch)
* Flake8 2.6.2 cause the ValueError from above without the patch and work with the patch
* Flake8 3.0.0b2 causes no error but doesn't actually checks any lines

So the question is: Do you care about supporting 2.5.0 and below? Otherwise #38 seems like one fix, but #31 with a bit of reworking (aka also importing pep8 when pycodestyle is not present) as I mentioned there is also viable.

Also this is further more verification that there are two issues and #31 or #38 only fix any versions of 2.6.0 or later as they use pycodestyle.

**Comment by twolfson** [comment]
Please move all code related discussions to the corresponding PR. I am tired of trying to match multiple conversations to the same issue.

**Comment by twolfson** [comment]
This has been repaired by @xZise in #38 and released in `0.6.0`
